mockgcp/common/httpmux/json.go

```diff
@@ -0,0 +1,59 @@
+package httpmux
+
+import (
+	"strings"
+
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"k8s.io/klog/v2"
+)
+
+func MarshalAsJSON(obj proto.Message) ([]byte, error) {
+	return protojson.MarshalOptions{Resolver: &protoResolver{}}.Marshal(obj)
+}
+
+type protoResolver struct {
+}
+
+var _ protoregistry.ExtensionTypeResolver = &protoResolver{}
+
+func (r *protoResolver) FindExtensionByName(message protoreflect.FullName) (protoreflect.ExtensionType, error) {
+	return protoregistry.GlobalTypes.FindExtensionByName(r.remapName(message))
+}
+
+func (r *protoResolver) FindExtensionByNumber(message protoreflect.FullName, field protoreflect.FieldNumber) (protoreflect.ExtensionType, error) {
+	return protoregistry.GlobalTypes.FindExtensionByNumber(r.remapName(message), field)
+}
+
+var _ protoregistry.MessageTypeResolver = &protoResolver{}
+
+func (r *protoResolver) FindMessageByName(message protoreflect.FullName) (protoreflect.MessageType, error) {
+	return protoregistry.GlobalTypes.FindMessageByName(r.remapName(message))
+}
+
+func (r *protoResolver) FindMessageByURL(url string) (protoreflect.MessageType, error) {
+	if strings.HasPrefix(url, "type.googleapis.com/google.") {
+		s := "type.googleapis.com/mockgcp." + strings.TrimPrefix(url, "type.googleapis.com/google.")
```

Comment:
Is the intent for swapping google. for mockgcp. (when we later swap it back) to ensure that we don't accidentally send to a live system during the test? I think it would be good to add a comment about how/why we temporarily make this change.

Reply:
I'll add a comment. This is a workaround for proto oddities, where you can't load two proto messages with the same proto name but different go packages.  (Now I think about it, that's probably precisely to support Any, which is what we're dealing with here!)